kmath-core/src/commonMain/kotlin/space/kscience/kmath/linear/LinearSpace.kt

```diff
@@ -0,0 +1,204 @@
+package space.kscience.kmath.linear
+
+import space.kscience.kmath.misc.UnstableKMathAPI
+import space.kscience.kmath.nd.*
+import space.kscience.kmath.operations.*
+import space.kscience.kmath.structures.Buffer
+import space.kscience.kmath.structures.BufferFactory
+import kotlin.reflect.KClass
+
+/**
+ * Alias for [Structure2D] with more familiar name.
+ *
+ * @param T the type of items.
+ */
+public typealias Matrix<T> = Structure2D<T>
+
+/**
+ * Alias or using [Buffer] as a point/vector in a many-dimensional space.
+ */
+public typealias Point<T> = Buffer<T>
+
+/**
+ * Basic operations on matrices and vectors. Operates on [Matrix].
+ *
+ * @param T the type of items in the matrices.
+ * @param M the type of operated matrices.
+ */
+public interface LinearSpace<T : Any, out A : Ring<T>> {
+    public val elementAlgebra: A
+
+    /**
+     * Produces a matrix with this context and given dimensions.
+     */
+    public fun buildMatrix(rows: Int, columns: Int, initializer: A.(i: Int, j: Int) -> T): Matrix<T>
+
+    /**
+     * Produces a point compatible with matrix space (and possibly optimized for it).
+     */
+    public fun buildVector(size: Int, initializer: A.(Int) -> T): Point<T>
+
+    public operator fun Matrix<T>.unaryMinus(): Matrix<T> = buildMatrix(rowNum, colNum) { i, j ->
+        -get(i, j)
+    }
+
+    public operator fun Point<T>.unaryMinus(): Point<T> = buildVector(size) {
+        -get(it)
+    }
+
+    /**
+     * Matrix sum
+     */
+    public operator fun Matrix<T>.plus(other: Matrix<T>): Matrix<T> = buildMatrix(rowNum, colNum) { i, j ->
+        get(i, j) + other[i, j]
+    }
+
+
+    /**
+     * Vector sum
+     */
+    public operator fun Point<T>.plus(other: Point<T>): Point<T> = buildVector(size) {
+        get(it) + other[it]
+    }
+
+    /**
+     * Matrix subtraction
+     */
+    public operator fun Matrix<T>.minus(other: Matrix<T>): Matrix<T> = buildMatrix(rowNum, colNum) { i, j ->
+        get(i, j) - other[i, j]
+    }
+
+    /**
+     * Vector subtraction
+     */
+    public operator fun Point<T>.minus(other: Point<T>): Point<T> = buildVector(size) {
+        get(it) - other[it]
+    }
+
+
+    /**
+     * Computes the dot product of this matrix and another one.
+     *
+     * @receiver the multiplicand.
+     * @param other the multiplier.
+     * @return the dot product.
+     */
+    public infix fun Matrix<T>.dot(other: Matrix<T>): Matrix<T> {
+        require(colNum == other.rowNum) { "Matrix dot operation dimension mismatch: ($rowNum, $colNum) x (${other.rowNum}, ${other.colNum})" }
+        return elementAlgebra {
+            buildMatrix(rowNum, other.colNum) { i, j ->
+                var res = zero
+                for (l in 0 until colNum) {
```

Comment:
Replace with (0 until colNum).sum { ... }?

Reply:
Nope, it would create an additional iterator. for-in has an intrinsic to avoid that.